File: src/series.go
// series.go
//
// Copyright (c) 2026 PDFjet Software
// Licensed under the MIT License. See LICENSE file in the project root.

package pdfjet

import (
	"github.com/edragoev1/pdfjet/v9/src/shape"
)

// Series is one series of a Chart: its points, the line that connects them
// when it is drawn, and the marker of the points added by their coordinates.
// A series is created with Chart.AddSeries; its name is listed in the legend
// of the chart. See Example_09.
type Series struct {
	name   string
	points []*Point

	strokeColor       [3]float32
	hasStrokeColor    bool // false: the next color of the palette
	strokeWidth       float32
	strokeDashPattern string
	drawPath          bool

	shape  shape.Shape
	radius float32
}

// newSeries creates a series with the default line and marker.
func newSeries(name string) *Series {
	series := new(Series)
	series.name = name
	series.points = make([]*Point, 0)
	series.strokeWidth = 1.0
	series.strokeDashPattern = "[] 0"
	series.shape = shape.Circle
	series.radius = 2.0
	return series
}

// AddPoint adds a point with the marker of this series.
// @param x the x value.
// @param y the y value.
func (series *Series) AddPoint(x, y float32) *Series {
	series.points = append(series.points, NewPoint(x, y).SetShape(series.shape).SetRadius(series.radius))
	return series
}

// AddPointWithMarker adds a point with its own marker: its shape, radius and
// colors. A point without a stroke color is drawn in the color of the series.
func (series *Series) AddPointWithMarker(point *Point) *Series {
	series.points = append(series.points, point)
	return series
}

// SetDrawPath sets whether the points are connected with a line, in the order
// they were added. The default is false: only the markers are drawn.
func (series *Series) SetDrawPath(drawPath bool) *Series {
	series.drawPath = drawPath
	return series
}

// SetStrokeColor sets the color of the line and of the markers that have no
// color of their own. Without it the series has the next color of the palette.
// @param color the color as a 0xRRGGBB value, for example color.Blue.
func (series *Series) SetStrokeColor(color int32) *Series {
	r := float32((color>>16)&0xff) / 255.0
	g := float32((color>>8)&0xff) / 255.0
	b := float32((color)&0xff) / 255.0
	series.strokeColor = [3]float32{r, g, b}
	series.hasStrokeColor = true
	return series
}

// SetStrokeWidth sets the width of the line. The default is 1.
func (series *Series) SetStrokeWidth(strokeWidth float32) *Series {
	series.strokeWidth = strokeWidth
	return series
}

// SetStrokeDashPattern sets the dash pattern of the line, for example
// "[3 3] 0". The default is a solid line.
func (series *Series) SetStrokeDashPattern(strokeDashPattern string) *Series {
	series.strokeDashPattern = strokeDashPattern
	return series
}

// SetShape sets the shape of the marker of the points added by their
// coordinates. The default is shape.Circle; shape.Invisible draws no markers.
func (series *Series) SetShape(markerShape shape.Shape) *Series {
	series.shape = markerShape
	return series
}

// SetRadius sets the radius of the marker of the points added by their
// coordinates. The default is 2.
func (series *Series) SetRadius(radius float32) *Series {
	series.radius = radius
	return series
}

File: Sources/PDFjet/Series.swift
/**
 * Series.swift
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */

///
/// One series of a Chart: its points, the line that connects them when it is
/// drawn, and the marker of the points added by their coordinates. A series is
/// created with Chart.addSeries; its name is listed in the legend of the chart.
/// See Example_09.
///
public class Series {
    let name: String
    var points = [Point]()
    // True for a point added by its coordinates, drawn with the marker of the series
    var seriesMarker = [Bool]()

    var strokeColor: [Float]?       // nil: the next color of the palette
    var strokeWidth: Float = 1.0
    var strokeDashPattern = "[] 0"
    var drawPath = false

    var shape = Shape.CIRCLE
    var radius: Float = 2.0

    init(_ name: String) {
        self.name = name
    }

    ///
    /// Adds a point with the marker of this series.
    ///
    /// - Parameter x: the x value.
    /// - Parameter y: the y value.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func addPoint(_ x: Float, _ y: Float) -> Series {
        points.append(Point(x, y))
        seriesMarker.append(true)
        return self
    }

    ///
    /// Adds a point with its own marker: its shape, radius and colors.
    /// A point without a stroke color is drawn in the color of the series.
    ///
    /// - Parameter point: the point.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func addPoint(_ point: Point) -> Series {
        points.append(point)
        seriesMarker.append(false)
        return self
    }

    ///
    /// Sets whether the points are connected with a line, in the order they
    /// were added. The default is false: only the markers are drawn.
    ///
    /// - Parameter drawPath: true to draw the line.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setDrawPath(_ drawPath: Bool) -> Series {
        self.drawPath = drawPath
        return self
    }

    ///
    /// Sets the color of the line and of the markers that have no color of
    /// their own. Without it the series has the next color of the palette.
    ///
    /// - Parameter color: the color as a 0xRRGGBB value, for example Color.blue.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setStrokeColor(_ color: Int32) -> Series {
        let r = Float((color >> 16) & 0xff)/255.0
        let g = Float((color >>  8) & 0xff)/255.0
        let b = Float((color)       & 0xff)/255.0
        self.strokeColor = [r, g, b]
        return self
    }

    ///
    /// Sets the width of the line. The default is 1.
    ///
    /// - Parameter strokeWidth: the line width.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setStrokeWidth(_ strokeWidth: Float) -> Series {
        self.strokeWidth = strokeWidth
        return self
    }

    ///
    /// Sets the dash pattern of the line, for example "[3 3] 0". The default
    /// is a solid line.
    ///
    /// - Parameter strokeDashPattern: the dash pattern.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setStrokeDashPattern(_ strokeDashPattern: String) -> Series {
        self.strokeDashPattern = strokeDashPattern
        return self
    }

    ///
    /// Sets the shape of the marker of the points added by their coordinates.
    /// The default is Shape.CIRCLE; Shape.INVISIBLE draws no markers.
    ///
    /// - Parameter shape: the shape.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setShape(_ shape: Shape) -> Series {
        self.shape = shape
        return self
    }

    ///
    /// Sets the radius of the marker of the points added by their coordinates.
    /// The default is 2.
    ///
    /// - Parameter radius: the radius.
    /// - Returns: this Series object.
    ///
    @discardableResult
    public func setRadius(_ radius: Float) -> Series {
        self.radius = radius
        return self
    }
}   // End of Series.swift
